### ISSUE TITLE
image_molecules is not working for some of our trajectories

### ISSUE THREAD

**Issue by zhang-ivy** [mdtraj/mdtraj#1625]
@glass-w and I are having trouble imaging the attached trajectory. This trajectory is made up of 2 protein chains, 2 glycan chains, 1 chain for structural ions, and 1 chain for solvent ions. It looks like the two protein chains (green and yellow) are split across the box.

![Screen Shot 2021-02-08 at 3 19 38 PM](https://user-images.githubusercontent.com/35546250/107276502-1a248c80-6a21-11eb-82a3-e14b021bb547.png)

Here is our code:
```
import mdtraj as md
solute = md.load("equilibrated_for_imaging.pdb")
traj = md.load('positions.xtc', top=solute)
anchors = set(solute.atom_slice(solute.topology.select('protein')).topology.atoms)
traj.image_molecules(anchor_molecules=[anchors])
traj.save_dcd("test.dcd")
```

[mdtraj_imaging_example.zip](https://github.com/mdtraj/mdtraj/files/5946588/mdtraj_imaging_example.zip)

**Comment by sroet** [comment]
**TLDR: this should work**
```python
import mdtraj as md
solute = md.load("equilibrated_for_imaging.pdb")
traj = md.load('positions.xtc', top=solute)

#Mimic https://github.com/mdtraj/mdtraj/blob/e968df9350723b387a86aeced41a30f0a62a2dc9/mdtraj/core/topology.py#L1237
all_molecules = traj.topology.find_molecules()
all_molecules.sort(key=lambda x: -len(x))
# grab the 2 biggest molecules
anchors2 = all_molecules[:2]
traj.image_molecules(anchor_molecules=anchors2)
traj.save_dcd("test2.dcd")
```

**Why your code did not do what you expect:** 
`image_molecules` loops through all molecules in the topology and tests if they are in the list of `anchor_molecules`. This is done by a `set` equal comparison. 
What you gave as `anchors` is a set of atoms that contains more than 1 molecule,  so none of the molecules in the topology matches that set, and therefor none of them are being selected as anchors.

The code above mirrors https://github.com/mdtraj/mdtraj/blob/e968df9350723b387a86aeced41a30f0a62a2dc9/mdtraj/core/topology.py#L1237 

It grabs all molecules, sorts them based on size and then grab the first two because we know we only want to center on the biggest 2 molecules. It essentially generates a list in the shape `[set(chain1_atoms), set(chain2_atoms)]` and uses that as the anchor list. 

As you know your system, you can probably create a similar thing by doing two smart `topology.select` call instead of 1.

traj after old code:
![Screenshot from 2021-02-09 09-00-21](https://user-images.githubusercontent.com/19587267/107333142-5a354f00-6ab5-11eb-89cf-d8240940eb70.png)

traj after new code:
![Screenshot from 2021-02-09 09-00-43](https://user-images.githubusercontent.com/19587267/107333168-60c3c680-6ab5-11eb-9c41-d2b4e2fd2d26.png)

**Comment by glass-w** [comment]
@sroet thanks for your snippet! That has fixed the issue and the chains are now together. I tried to extend / adapt this to specify my own topology so I didn't have to use `find_molecules()` and instead parse the protein chains manually. However, when trying this the trajectory wasn't fixed, even thoguh both anchor selections should be the same as in your original example. 

I compared the `set` selections returned and noticed that for the larger chain they are not equal:

```
import mdtraj as md

solute = md.load("equilibrated_for_imaging.pdb")
all_molecules = solute_trajectory.topology.find_molecules() 
all_molecules.sort(key=lambda x: -len(x))
anchors = all_molecules[:2]

chain2_large = set(solute.atom_slice(solute.topology.select("chainid 2")).topology.atoms) # the larger chain
chain0_small = set(solute.atom_slice(solute.topology.select("chainid 0")).topology.atoms) # the smaller chain

chain2_large == anchors[0] # this should be True
False
chain2_small == anchors[1]
True

anchors = [chain2_large, chain0_small]

traj = md.load('positions.xtc', top=solute)
traj.image_molecules(anchor_molecules=anchors)
traj.save_dcd("test_new.dcd")
```

I'm confused as to why these should be different, if I understand correctly `chain2_large` should match `anchors[0]`.
The only difference in generating them is between `find_molecules()` and `atom_slice()`, where the former returns a list of sets whilst the latter returns a generator obejct (that is then converted to a set in the snippet above). I've attached a screenshot of the trajectory from the snippet:

![Screenshot 2021-02-09 at 20 35 02](https://user-images.githubusercontent.com/18193462/107425004-4cfc7c80-6b16-11eb-80dc-8216a3acb513.png)

**Comment by zhang-ivy** [comment]
Thank you so much @sroet -- your detailed response really helped me better understand how `image_molecules()` works.
I've investigated @glass-w's problem above, and found that the reason why this
```
chain2_large == anchors[0] # this should be True
False
```
is happening. It's because when we generate the atom set using this line,
```
chain2_large = set(solute.atom_slice(solute.topology.select("chainid 2")).topology.atoms) # the larger chain
```
, the `atom_slice()` call creates a new trajectory with the desired subset of atoms, but these atoms are renumbered started from 0 (whereas the atom set returned by `find_molecules()` is not renumbered starting from 0). To replace this line, we can use:
```
chain2_large = set()
desired_indices = solute.topology.select("chainid 2")
for atom in solute.topology.atoms:
    if atom.index in desired_indices:
        chain2_large.add(atom)
```

**Comment by zhang-ivy** [comment]
However, when we apply a similar code snippet to a different system (with 3 protein chains instead of 2), we are still seeing imaging problems for some of the frames. 

Here is the code:
```
import mdtraj as md
solute = md.load("equilibrated_for_imaging.pdb")
traj = md.load('positions.xtc', top=solute)

all_molecules = solute.topology.find_molecules()
all_molecules.sort(key=lambda x: -len(x))
anchors = all_molecules[:3] # Grab the 3 largest chains, because now there are 3 protein chains

traj = traj.image_molecules(anchor_molecules=anchors)
traj.save_dcd("test.dcd")
```
We also tried using only the 2 largest chains, but that didn't fix our issue either.

Here is a problematic frame. The protein chains are colored cyan, green, and magenta.
![Screen Shot 2021-02-09 at 6 05 14 PM](https://user-images.githubusercontent.com/35546250/107440285-5e875980-6b01-11eb-887f-f81cf15a4b5c.png)

[mdtraj_imaging_example_2.zip](https://github.com/mdtraj/mdtraj/files/5954764/mdtraj_imaging_example_2.zip)

**Comment by rmcgibbo** [comment]
I haven't looked into this, and probably wont have the chance to.

But FYI: In my organization, the way we often handle imaging in complicated cases in order to give chemists the flexibility to make it work when the default behavior doesn't work is to have the option to add "glue", which is like a fictitious bond between two atoms used _only_ for the purpose of determining connected-component-ness during imaging. So you might be able to modify your topology to temporarily glue your protein chains together or something and thus cause the imaging thing to do the right thing.

Not sure if that will work for your case, but it might.

**Comment by sroet** [comment]
> However, when we apply your recommended code snippet to a different system (with 3 protein chains instead of 2), we are still seeing imaging problems for some of the frames.

I just looked at the frame and what you are seeing is the fact that there is more than one "correct" solution. Your proteins span the whole box vector, so the algorithm tries to put as many anchor atoms as possible into the box. That is "a" correct and reasonable solution, but can lead to these broken representations. What solution did you expect?

Options I quickly ran through:
- select the 3 biggest, sees the "issue" you are describing
- select the 2 biggest: keeps cyan and green together, but breaks magenta/yellow molecule 
- select the 1st and 3rd biggest:  keeps cyan and green together, but breaks magenta/yellow molecule
- select the 3rd and 4th biggest; keeps magenta/yellow together and centered and cyan/green together but sticking out of the box (at least for the 4 frames you supplied) (this might be the solution you want?)

Again, I would like to stress that none of these are "wrong" answers, they are all very suitable answers as it needs to to pick "a" point where to wrap the boxvectors and your system spans the whole box. You might however look for a slightly different point.

Here is the code I used to check (I run this inside jupyter and uses [`nglview`](http://nglviewer.org/nglview/latest/) to visualise the trajectory inside the notebook)
```python
import mdtraj as md
solute = md.load("equilibrated_for_imaging.pdb")
traj = md.load('positions.xtc', top=solute)

#Mimic https://github.com/mdtraj/mdtraj/blob/e968df9350723b387a86aeced41a30f0a62a2dc9/mdtraj/core/topology.py#L1237
all_molecules = traj.topology.find_molecules()
all_molecules.sort(key=lambda x: -len(x))
# grab the 2 bigest molecules
anchors2 = [all_molecules[2],all_molecules[3]]
traj.image_molecules(anchor_molecules=anchors2[:])
traj.save_dcd("test2.dcd")

import nglview
view = nglview.show_mdtraj(traj)
view
```

**Comment by zhang-ivy** [comment]
Thank you both for your help! 
> just looked at the frame and what you are seeing is the fact that there is more than one "correct" solution
@sroet this makes sense -- we looked more carefully at our frames and realized that our box was not appropriately shaped/sized, causing the protein to rotate and interact with its image. We are re-solvating with a different box and going to re-run our simulations.
